### PR TITLE
implementing elastic search service

### PR DESCRIPTION
**Pull request by kardi31**

### CODE REVIEW

File: Builder/ElasticDocumentBuilder.php [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+
+    /**
+     * @param array $inputData
+     */
+    public function addData(array $inputData)
+    {
+        if (isset($this->data['body'])) {
+            $this->data['body'] = array_merge($this->data['body'], $inputData);
+        } else {
+            $this->data['body'] = $inputData;
```
Comment (gsdevme) [inline Builder/ElasticDocumentBuilder.php]: Seems to me we could just do `$this->data['body'] = []` in the construct then remove the need for condition completely.

File: Builder/ElasticDocumentBuilder.php [hunk 2]
```diff
@@ -0,0 +1,105 @@
+<?php
+
+namespace Markup\NeedleBundle\Builder;
+
+class ElasticDocumentBuilder implements ElasticDocumentInterface
+{
+    /**
+     * @var array
+     */
+    protected $data = [];
```
Comment (gsdevme) [inline Builder/ElasticDocumentBuilder.php]: You should set this in the constructor really.

File: Service/ElasticSearchService.php [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+    }
+
+    public function getDocument($params)
+    {
+        return $this->getElasticQuery()->get($params);
+    }
+
+    public function search($params)
+    {
+//        dump($params);exit;
```
Comment (gsdevme) [inline Service/ElasticSearchService.php]: 💩

File: Filter/Field/TextFilterField.php [hunk 4]
```diff
@@ -0,0 +1,42 @@
+<?php
+// https://www.elastic.co/guide/en/elasticsearch/client/php-api/current/_search_operations.html
```
Comment (gsdevme) [inline Filter/Field/TextFilterField.php]: If this provides useful reference you should add it to the class docblocks, this isn't really a supported way to provide documentation

File: Builder/ElasticDocumentBuilder.php [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+     */
+    public function addMapping($fieldName, $fieldType, ?array $overrideValues = null)
+    {
+        switch ($fieldType):
+            case self::FIELD_TYPE_DATE:
+                $mapping = $this->addMappingForDate($fieldName, $overrideValues);
+                break;
+            default:
+                return;
+        endswitch;
```
Comment (gsdevme) [inline Builder/ElasticDocumentBuilder.php]: You should use braces here

File: Builder/ElasticSearchBuilder.php [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+        array_push($this->existTypeFields, $fieldName);
+    }
+
+    /**
+     * @param string $fieldName
+     * @return bool
+     */
+    public function isRangeType(string $fieldName)
+    {
+        if (in_array($fieldName, $this->rangeTypeFields)) {
```
Comment (gsdevme) [inline Builder/ElasticSearchBuilder.php]: You can just `return in_array($fieldName, $this->rangeTypeFields);`

File: Builder/ElasticSearchBuilder.php [hunk 7]
```diff
@@ -0,0 +1,206 @@
...
+        return false;
+    }
+
+    /**
+     * @param string $fieldName
+     * @return bool
+     */
+    public function isArrayType(string $fieldName)
+    {
+        if (in_array($fieldName, $this->arrayTypeFields)) {
```
Comment (gsdevme) [inline Builder/ElasticSearchBuilder.php]: same here

File: Builder/ElasticSearchBuilder.php [hunk 8]
```diff
@@ -0,0 +1,206 @@
...
+        return false;
+    }
+
+    /**
+     * @param string $fieldName
+     * @return bool
+     */
+    public function isExistType(string $fieldName)
+    {
+        if (in_array($fieldName, $this->existTypeFields)) {
```
Comment (gsdevme) [inline Builder/ElasticSearchBuilder.php]: same here

File: Builder/ElasticSearchBuilder.php [hunk 9]
```diff
@@ -0,0 +1,206 @@
+<?php
+
+namespace Markup\NeedleBundle\Builder;
+
+use Markup\NeedleBundle\Filter\Field\ArrayFilterField;
+use Markup\NeedleBundle\Filter\Field\ExistsFilterField;
+use Markup\NeedleBundle\Filter\Field\RangeFilterField;
+use Markup\NeedleBundle\Filter\Field\TextFilterField;
+
+class ElasticSearchBuilder implements ElasticDocumentInterface
```
Comment (gsdevme) [inline Builder/ElasticSearchBuilder.php]: As a concept I think this shouldn't be called `Builder`.. its also using `arrays` so much its hard to see value in it, its not really a like for like with the `SolariumSelectQueryBuilder` so Im having to hard time understanding why it exists in Needle.. if Needle is at least meant to provide functionality across both platforms.

File: Service/ElasticSearchService.php [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+    }
+
+    public function putMapping($params)
+    {
+        return $this->getElasticQuery()->indices()->putMapping($params);
+    }
+
+    private function getElasticQuery()
+    {
+        return $this->elastic::create()->build();
```
Comment (gsdevme) [inline Service/ElasticSearchService.php]: is there no configuration this? Its unlikely we will have PHP & Elastic search on the same server after all.

File: Service/ElasticSearchService.php [hunk 11]
```diff
@@ -0,0 +1,89 @@
+<?php
+
+namespace Markup\NeedleBundle\Service;
+
+use Doctrine\Common\Collections\ArrayCollection;
```
Comment (gsdevme) [inline Service/ElasticSearchService.php]: Im not sure all these imports/use statements are used 🤔

File: Filter/Field/RangeFilterField.php [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+     * @var string|null
+     */
+    private $minValue;
+
+    /**
+     * @var string|null
+     */
+    private $maxValue;
+
+    public function __construct(string $fieldName, array $value)
```
Comment (gsdevme) [inline Filter/Field/RangeFilterField.php]: `$value` is too generic.. I have no clue how to construct this object without looking at it.. this should likely be `\DateTime $start, \DateTime $end` surely?

**Comment by shieldo** [comment]
Yes, this actually needs proper discussion before embarking on it. The whole purpose of the Needle library is to provide a generic abstraction in front of different backends: Solr, Elasticsearch, Algolia, whatever else. This seems to completely sidestep that, so it should really be its own library if it serves a purpose over and above the Elastica library.

**Comment by shieldo** [comment]
Closing as this has been implemented separately/ via the Needle abstraction.

**Pull request closed without merging**